File: LinkedList/python/leetcode24_Swap_Nodes_in_Pairs.py
# Given a linked list, swap every two adjacent nodes and return its head.

# Example:

# Given 1->2->3->4, you should return the list as 2->1->4->3.



# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = heads = ListNode(0)
        heads.next = head
        while heads.next and heads.next.next:
            heads.next,heads.next.next,c = heads.next.next, heads.next, heads.next.next.next
            heads.next.next.next = c
            heads = heads.next.next
        return dummy.next

# Time: O(n)
# Space: O(1)
# Difficulty: medium

File: LinkedList/java/leetcode24_Swap_Nodes_in_Pairs.java
//Given a linked list, swap every two adjacent nodes and return its head.
//
//Example:
//
//Given 1->2->3->4, you should return the list as 2->1->4->3.

package wenxinjie;

public class leetcode24_Swap_Nodes_in_Pairs {
	public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0);
        ListNode head1 = dummy;
        head1.next = head;
        while (head1.next != null && head1.next.next != null){
            ListNode c = head1.next.next.next, b = head1.next.next, a = head1.next;
            head1.next = b; 
            head1.next.next = a; 
            head1.next.next.next = c;
            head1 = head1.next.next;
        }
        return dummy.next;       
    }
}

//Time: O(n)
//Space: O(1)
//Difficulty: medium